From 2f636f1fb6011d77e3f910f5c2388747715f7d6a Mon Sep 17 00:00:00 2001
From: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
Date: Fri, 5 Apr 2019 17:26:09 +0530
Subject: [PATCH 25/47] Add option to activate alternative 'coda' VPU driver.

Freescale's normally active VPU DT node &vpu_fsl sets compatible =
"fsl,imx6-vpu", which leads to drivers/mxc/vpu/mxc_vpu.c.

But imx6qdl.dtsi contains an alternative definition &vpu for the exact same
VPU device (same base address!), tagging it "fsl,imx6q/dl-vpu", which
invokes a different driver in drivers/media/platform/coda/*.

Freescale's defconfig compiles in (=y) both, but the &vpu node is disabled
and no .dts I could find ever changes that. Only the mxc_vpu driver is used.
But this driver's source is only 30 kB and hasn't changed since kernel 3.10,
whereas the coda driver has grown lots and is now at 150 kB. Maybe Freescale
intends to switch over but isn't quite there yet? Or the coda driver is used
within Yocto and some recipe trickery, which I couldn't find, activates it?
I saw that the coda sources request some firmware binaries, maybe that's why
it's not used per default?

Anyway, this patch adds a commented option allowing to play with this thing.
Our defconfig leaves CONFIG_MXC_VPU=y, but downgrades CONFIG_VIDEO_CODA=m.

Incidentally there are also two DRM drivers. CONFIG_DRM_VIVANTE is the one
used by Freescale's (and our) defconfig, whereas CONFIG_DRM_IMX is in
mainline. The latter depends on CONFIG_IMX_IPUV3_CORE, which collides with
CONFIG_MXC_IPU_V3 (needed by CONFIG_FB_MXC): both IPU core drivers apply to
the same DT node and define conflicting symbols, so we can't compile them
both. Another area with duplicate drivers is I2S audio, see
CONFIG_SND_SOC_FSL_ASOC_CARD.

Signed-off-by: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
---
 arch/arm/boot/dts/lec-imx6.dtsi | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 8a5c17e..92789f3 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -19,6 +19,14 @@
 /* The LEC-Base carrier doesn't include cameras, but allows connecting some. */
 #define CONFIG_SER_CAMERA	3c /* I2C addr of serial camera, hex without "0x" */
 //#define CONFIG_PAR_CAMERA	3c /* I2C addr of parallel camera */
+
+/* Freescale defines 2 alternative DT nodes for the VPU: The one normally active
+ * loads a driver called 'mxc_vpu'. The other, disabled one, would invoke a
+ * different driver named 'coda'. I don't know what state it is in; none of
+ * Freescale's sample dts files enable it, but the code is there, appearently
+ * dormant, in imx6qdl.dtsi. 'Coda' requires a v4l-coda960-imx6*.bin firmware. */
+#define VPU_CODA	0
+
 #include <dt-bindings/clock/imx6qdl-clock.h>
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
@@ -334,6 +342,14 @@
 	status = "okay";
 };
 
+#if VPU_CODA
+&vpu {
+	status = "okay";
+};
+&vpu_fsl {
+	status = "disabled";
+};
+#endif
 
 #if CONFIG_I2S_AUDIO
 &audmux {	/* SMARC I2S0 (instead of SPI0) */
-- 
2.7.4

